src/API/Management/Users.php

```diff
@@ -1,88 +1,262 @@
+    {
+        if (empty($data['connection'])) {
+            throw new \Exception('Missing required "connection" field.');
+        }
 
-        return $this->apiClient->post()
-            ->users()
-            ->withHeader(new ContentType('application/json'))
+        if ('sms' === $data['connection'] && empty($data['phone_number'])) {
+            // "phone_number" field is required for an sms connection.
+            throw new \Exception('Missing required "phone_number" field for sms connection.');
+        } elseif ('sms' !== $data['connection']) {
+            // "email" field is required for email and DB connections.
+            if (empty($data['email'])) {
+                throw new \Exception('Missing required "email" field.');
+            }
+
+            // Passwords are required for DB connections.
+            if ('email' !== $data['connection'] && empty($data['password'])) {
+                throw new \Exception('Missing required "password" field for DB connection.');
+            }
+        }
+
+        return $this->apiClient->method('post')
+            ->addPath('users')
             ->withBody(json_encode($data))
             ->call();
     }
 
-    public function getAll($params = array()) {
-        return $this->search($params);
-    }
-
-    public function search($params = array()) {
+    /**
+     * Search all Users.
+     *
+     * @param array $params - Search parameters to send:
+     *      - "fields", "include_fields", "page", and "per_page" keys here will override the explicit parameters.
+     *      - Queries using "search_engine" set to "v2" should be migrated to v3; see migration @link below
+     * @param null|string|array $fields - Fields to include or exclude from the result, empty to retrieve all fields:
+     *      - Including only the fields required can speed up API calls significantly.
+     *      - Arrays will be converted to comma-separated strings
+     * @param null|boolean $include_fields - True to include $fields, false to exclude $fields.
+     * @param null|integer $page - Page number to get, zero-based.
+     * @param null|integer $per_page - Number of results to get, null to return the default number.
+     *
+     * @return mixed
+     *
+     * @throws \Exception
+     *
+     * @link https://auth0.com/docs/users/search/v3#migrate-from-search-engine-v2-to-v3
+     * @link https://auth0.com/docs/api/management/v2#!/Users/get_users
+     * @link https://auth0.com/docs/users/search/best-practices
+     * @link https://auth0.com/docs/users/search/v3/query-syntax
+     * @link https://auth0.com/docs/users/search/v2/query-syntax
+     */
+    public function getAll($params = [], $fields = null, $include_fields = null, $page = null, $per_page = null)
+    {
+        $params = is_array($params) ? $params : [];
 
-        $client = $this->apiClient->get()
-            ->users();
+        // Fields to include/exclude.
+        if (!isset($params['fields']) && null !== $fields) {
```

Comment:
I'm not following your suggestion here. This check is to make sure we don't override existing parameters set (something that's not necessary on the other endpoints with how they are built).

Reply:
that's just what I mean. You have independent params for `field`, `include_fields`, `page` and `per_page`. Those should prevail over the `params` dictionary. So, code should just:

```
$final_params = $params
if ($page != null){
  $final_params["page"] = $page
}

//and so on with the rest
```

Reply:
It's a little confusing either way, IMO, because it's handling a situation where someone is setting, say, `fields` in two places. I think someone is going to have some debugging to do either way.

Reply:
no debugging is needed as long as it's well documented. Being one case or the other. Both are corrects. I like more the one I suggested in my previous comment since the previous signature didn't have anything but the _params_ dictionary, so anything added later (talking about order) would replace them, but I accept yours as well 👍  Choose whatever that sticks to the other entities' methods behavior.